### PR TITLE
Add a code sample for item-item recommendation using BigQuery ML matrix factorization and ScaNN

### PR DESCRIPTION
**Pull request by ksalama**
This PR contains a sample code for training and serving embeddings for real-time similarity matching. The system utilizes BigQuery ML Matrix Factorization model to train the embeddings, and the open-source ScaNN framework to build and approximate nearest neighbour index.

1. Compute pointwise mutual information (PMI) between items based on their cooccurrences.
2. Train item embeddings using BigQuery ML Matrix Factorization, using item PMI as implicit feedback.
3. Export and post-process the embeddings from BigQuery ML model to Cloud Storage as CSV files using Cloud Dataflow.
4. Implement an embedding lookup model using Keras and deploys it to AI Platform Prediction.
5. Serve the embedding as an approximate nearest neighbor index using ScaNN

### CODE REVIEW

**Comment by ksalama** [comment]
@polong-lin -  Kindly review the PR

File: retail/recommendation-system/bqml-scann/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Real-time Item-to-item Recommendation BigQueryML Matrix Factorization and ScaNN
```
Comment (polong-lin) [inline retail/recommendation-system/bqml-scann/README.md:1]: add space: BigQuery ML

File: retail/recommendation-system/bqml-scann/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+This directory contains a sample code for training and serving embeddings 
+for real-time similarity matching. The system utilizes [BigQuery ML Matrix Factorization](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-create-matrix-factorization)
+model to train the embeddings, and the open-source [ScaNN framework](https://ai.googleblog.com/2020/07/announcing-scann-efficient-vector.html) to build and
+approximate nearest neighbour index.
+
+1. Compute pointwise mutual information (PMI) between items based on their cooccurrences.
+2. Train item embeddings using BigQuery ML Matrix Factorization, using item PMI as implicit feedback.
+3. Export and post-process the embeddings from BigQuery ML model to Cloud Storage as CSV files using Cloud Dataflow.
+4. Implement an embedding lookup model using Keras and deploys it to AI Platform Prediction.
```
Comment (polong-lin) [inline retail/recommendation-system/bqml-scann/README.md:11]: typo: deploy **it** to

File: retail/recommendation-system/bqml-scann/index_builder/builder/indexer.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import os
+import scann
+import tensorflow as tf
+import numpy as np
+import math
+import pickle
+
+METRIC = 'dot_product'
+DIMENSIONS_PER_BLOCK = 2
+ANTISOTROPIC_QUANTIZATION_THRESHOLD = 0.2
```
Comment (polong-lin) [inline retail/recommendation-system/bqml-scann/index_builder/builder/indexer.py:24]: Is this a typo? 
Current:
**ANTISO**TROPIC_QUANTIZATION_THRESHOLD
Suggestion:
**ANISO**TROPIC_QUANTIZATION_THRESHOLD

File: retail/recommendation-system/bqml-scann/index_builder/builder/indexer.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  data_size = embeddings.shape[0] 
+  if not num_leaves:
+    num_leaves = int(math.sqrt(data_size))
+    
+  print('Start building the ScaNN index...')
+  scann_builder = scann.scann_ops.builder(embeddings, NUM_NEIGHBOURS, METRIC).tree(
+    num_leaves=num_leaves, 
+    num_leaves_to_search=NUM_LEAVES_TO_SEARCH, 
+    training_sample_size=data_size).score_ah(
+      DIMENSIONS_PER_BLOCK, anisotropic_quantization_threshold=ANTISOTROPIC_QUANTIZATION_THRESHOLD).reorder(REORDER_NUM_NEIGHBOURS)
```
Comment (polong-lin) [inline retail/recommendation-system/bqml-scann/index_builder/builder/indexer.py:67]: Looks like a typo with the variable vs. parameter. aniso vs. antiso

File: retail/recommendation-system/bqml-scann/index_builder/builder/indexer.py (hunk 3 above)
Comment (ksalama) [reply retail/recommendation-system/bqml-scann/index_builder/builder/indexer.py]: Done

File: retail/recommendation-system/bqml-scann/index_builder/builder/indexer.py (hunk 4 above)
Comment (ksalama) [reply retail/recommendation-system/bqml-scann/index_builder/builder/indexer.py]: Done

File: retail/recommendation-system/bqml-scann/README.md (hunk 2 above)
Comment (ksalama) [reply retail/recommendation-system/bqml-scann/README.md]: Done

File: retail/recommendation-system/bqml-scann/README.md (hunk 1 above)
Comment (ksalama) [reply retail/recommendation-system/bqml-scann/README.md]: Done

**Pull request merged**